File: LeetCode/Problems/Python/200. Number of Islands.py
#BFS
#Time: O(m*n)
#Space: O(m*n)
class Solution:
    
    def bfs(self, curr, visited, grid):
        
        queue = [curr]
        directions = [[0,1],[0,-1],[1,0],[-1,0]]
        
        while queue:
            curr = queue.pop(0)
            
            for d in directions:
                if 0 <= curr[0] + d[0] < len(grid) and 0 <= curr[1]+d[1] < len(grid[0]):
                    next_pos = (curr[0] + d[0],curr[1] + d[1])
                    if grid[next_pos[0]][next_pos[1]] == "1" and next_pos not in visited:
                        visited[next_pos] = True
                        queue.append(next_pos)

        return visited
    
    def numIslands(self, grid: List[List[str]]) -> int:
        
        if not grid:
            return 0
        
        visited = {}
        ans = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == "1" and (i,j) not in visited:
                    ans +=1
                    
                    visited = self.bfs((i,j), visited, grid)

        return ans

#DFS
#Time: O(m*n)
#Space: O(m*n)
class Solution:
    def dfs(self, grid, row, col, visited):
        direction = [(0,1),(0,-1),(1,0),(-1,0)]
        stack = [(row,col)]
        while stack:
            r, c = stack.pop()
            for d in direction:
                new_r = r+d[0]
                new_c = c+d[1]
                if 0 <= new_r < len(grid) and 0 <= new_c < len(grid[0]):
                    if grid[new_r][new_c] == '1' and (new_r,new_c) not in visited:
                        visited[(new_r,new_c)] = True
                        stack.append((new_r,new_c))
                        
    def numIslands(self, grid: List[List[str]]) -> int:
        if grid is None:
            return 0
   
        visited = {}
        num_islands = 0
        for row in range(len(grid)):
            for col in range(len(grid[0])):
                if grid[row][col] == '1' and (row,col) not in visited:
                    num_islands+=1
                    visited[(row,col)] = True
                    self.dfs(grid, row, col, visited)
                    
                        
        return num_islands

File: LeetCode/Problems/C++/200. Number of Islands.cpp
class Solution {
public:
    void bfs(int i, int j, vector<vector<char>>& grid, int* visited, int M, int N){
        queue<pair<int, int>> neighbor;
        neighbor.push({i,j});
        int r, c;

        
        while(!neighbor.empty()){
            r = neighbor.front().first;
            c = neighbor.front().second;
            neighbor.pop();
            
            if(r-1 >= 0 && grid[r-1][c] == '1' && !visited[(r-1)*N+c]){
                visited[(r-1)*N+c] = 1;
                neighbor.push({r-1,c});
                }
            if(r+1 < M && grid[r+1][c] == '1' && !visited[(r+1)*N+c]){
                visited[(r+1)*N+c] = 1;
                neighbor.push({r+1,c});
                }
            if(c-1 >= 0 && grid[r][c-1] == '1' && !visited[r*N+c-1]){
                visited[r*N+c-1] = 1;
                neighbor.push({r,c-1});
                }
            if(c+1 < N && grid[r][c+1] == '1' && !visited[r*N+c+1]){
                visited[r*N+c+1] = 1;
                neighbor.push({r,c+1});
                }
        }
        
    }
    int numIslands(vector<vector<char>>& grid) {
        if(!grid.size()) return 0;
        
        int M = grid.size();
        int N = grid[0].size();
        int* visited = new int[M*N]();
        int island = 0;
        for(int i=0;i<M;i++){
            for(int j=0;j<N;j++){
                if(grid[i][j]=='1' && !visited[i*N + j]){
                    bfs(i,j, grid, visited, M, N);
                    island++;
                }
            }
        }
        return island;
    }
};